server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java

```diff
@@ -0,0 +1,140 @@
+import com.google.firebase.cloud.FirestoreClient;
+import com.google.sps.utils.Ad;
+import com.google.sps.utils.Constants;
+import java.io.FileInputStream;
+// import java.io.InputStream;
+// import java.util.ArrayList;
+// import java.util.Arrays;
+import java.util.HashMap;
+// import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Description: Class creates collection by state. Each state collection contains
+ *              two subcollections: 'advertisers', which contains information about
+ *              each advertiser's total ad spend, and 'stateData', which contains
+ *              information about each state's total ad spend.
+ * Author: Kira Toal
+ * Date: July 20, 2020
+ */ 
+public class StateSubcollectionBuilder {
+
+  private static final String PATH_TO_SERVICE_ACCOUNT = "./serviceAccountKey.json"; 
+  private static final String DATABASE_URL = "https://step9-2020-capstone.firebaseio.com"; 
+  private static final String MAIN_COLLECTION = "ads";
+  private static final String WRITE_COLLECTION = "dev_states";
+  private static final Set<String> GEO_TARGETS = Constants.TEST_GEO_TARGETS;
+  public static Firestore db; 
+
+  public static Firestore initializeApp() throws Exception {
+    // Set account and build options.
+    FileInputStream serviceAccount = new FileInputStream(PATH_TO_SERVICE_ACCOUNT);
+    FirebaseOptions options = new FirebaseOptions.Builder()
+        .setCredentials(GoogleCredentials.fromStream(serviceAccount))
+        .setDatabaseUrl(DATABASE_URL)
+        .build();
+    
+    // Initialize app.
+    if (FirebaseApp.getApps().isEmpty()) {
+      FirebaseApp.initializeApp(options);
+    }
+    db = FirestoreClient.getFirestore();
+    return db;
+  }
+
+  /**
+   * Maps advertisers to their total ad spend in a particular state
+   * @param documentsInState list of all document snapshots targeted
+   *                         at a particular state 
+   */
+    public static Map<String, Long> getAdvertiserToSpend(
+      List<QueryDocumentSnapshot> documentsInState) {
+    Map<String, Long> advertiserToSpend = new HashMap<>();
+
+    for (QueryDocumentSnapshot document: documentsInState) {
+      String advertiser = document.getString("advertiser");
+      long spendMax = document.getLong("spendMax");
+
+      if (advertiserToSpend.containsKey(advertiser)) {
+        spendMax += advertiserToSpend.get(advertiser); 
+      }
+      
+      advertiserToSpend.put(advertiser, spendMax);
+    }
+    
+    return advertiserToSpend;
+  }
+
+  /**
+   * Writes advertiser documents to Firestore
+   * @param state the current state subcollection (tells Firestore
+   *              where to write the advertiser documents)
+   * @param advertiserToSpend map that relates every advertiser
+   *              to their total ad spend in a particular state              
+   */
+  public static void updateStateSubcollection(String state,
```

Comment:
Are there tests for these methods?

Reply:
Not in this PR, but I'm currently working on one with tests and cloud functions to live update the states subcollections